src/algebra/group/units.lean

```diff
@@ -415,6 +415,10 @@ begin
   simp [h.unit_spec]
 end
 
+/-- `is_unit x` is decidable if we can decide if `x` comes from `Mˣ`. -/
+instance [monoid M] (x : M) : Π [decidable (∃ u : Mˣ, ↑u = x)], decidable (is_unit x) :=
+id
```

Comment:
In some cases, Lean fails to unify `x` with `id x` while matching two instances.
```suggestion
instance [monoid M] (x : M) [h : decidable (∃ u : Mˣ, ↑u = x)] : decidable (is_unit x) := h
```

Reply:
I was curious about this; we have this id pattern in a lot of places so I thought it might be ok.

I think you're probably right; but do you have a mwe of the failure?

Reply:
I have made the change, although I have to admit that I don't really understand the background.

Reply:
> I was curious about this; we have this id pattern in a lot of places so I thought it might be ok.
> 
> I think you're probably right; but do you have a mwe of the failure?

I remember that we had extra `id`s in pi instances and some unification was failing because of them. I don't have an MWE.